internal/app/machined/pkg/controllers/runtime/watchdog_timer.go

```diff
@@ -0,0 +1,135 @@
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
+			ID:        optional.Some(config.V1Alpha1ID),
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *WatchdogTimerController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *WatchdogTimerController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	ticker := time.NewTicker(3 * time.Second)
+	defer ticker.Stop()
+
+	var wd *os.File
+
+	wdClose := func() {
+		logger.Info("Closing hardware watchdog", zap.String("filename", wd.Name()))
+		// Magic close: make sure old watchdog won't trip after we close it
+		if _, err := wd.WriteString("V"); err != nil {
+			logger.Error("Failed to send magic close to watchdog", zap.String("filename", wd.Name()))
+		}
+
+		if err := wd.Close(); err != nil {
+			logger.Error("Failed to close watchdog", zap.String("filename", wd.Name()))
+		}
+
+		wd = nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			if wd != nil {
+				if _, _, err := syscall.Syscall(syscall.SYS_IOCTL, wd.Fd(), unix.WDIOC_KEEPALIVE, 0); err != 0 {
+					return fmt.Errorf("failed to feed watchdog: %w", err)
+				}
+			}
+
+			continue
+		case <-r.EventCh():
+		}
+
+		r.ResetRestartBackoff()
+
+		cfg, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.V1Alpha1ID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting watchdog config: %w", err)
+			}
+		}
+
+		if cfg != nil && cfg.Config().Machine() != nil {
+			dev := cfg.Config().Machine().Kernel().WatchdogDevice()
+			timeout := cfg.Config().Machine().Kernel().WatchdogTimeout()
+
+			if dev == "" || timeout == 0 {
+				continue
+			}
+
+			// Close the watchdog if requested to use new one
+			if wd != nil && wd.Name() != dev {
+				wdClose()
+			}
+
+			if wd == nil {
+				wd, err = os.OpenFile(dev, syscall.O_RDWR, 0o600)
+				if err != nil {
+					return errors.Errorf("failed to open watchdog device: %s", err)
+				}
+
+				logger.Info("Opened hardware watchdog", zap.String("filename", dev))
+				// TODO: support reboot/shutdown watchdogs
+				defer wdClose()
+			}
+
+			if _, _, err := syscall.Syscall(syscall.SYS_IOCTL, wd.Fd(), uintptr(unix.WDIOC_SETTIMEOUT), uintptr(unsafe.Pointer(&timeout))); err != 0 {
+				return fmt.Errorf("failed to set watchdog timeout: %w", err)
+			}
+
+			// 3 pings per timeout should suffice in any case
+			ticker = time.NewTicker(time.Duration(timeout/3) * time.Second)
+
+			if _, _, err := syscall.Syscall(syscall.SYS_IOCTL, wd.Fd(), uintptr(unix.WDIOC_KEEPALIVE), uintptr(0)); err != 0 {
+				return fmt.Errorf("failed to feed watchdog: %w", err)
+			}
+
+			logger.Info("Set hardware watchdog timeout", zap.Int("timeout", timeout), zap.Int("feed_interval", timeout/3))
```

Comment:
log messages should be lowercase 
```suggestion
			logger.Info("set hardware watchdog timeout", zap.Int("timeout", timeout), zap.Int("feed_interval", timeout/3))
```